nomad/alloc_endpoint.go

```diff
@@ -80,9 +81,40 @@ func (a *Alloc) GetAlloc(args *structs.AllocSpecificRequest,
 			if err != nil {
 				return err
 			}
-			out, err := snap.AllocByID(args.AllocID)
-			if err != nil {
-				return err
+
+			var out *structs.Allocation
+
+			// Exact lookup if the identifier length is 36 (full UUID)
+			if len(args.AllocID) == 36 {
+				out, err = snap.AllocByID(args.AllocID)
+				if err != nil {
+					return err
+				}
+			} else {
+				iter, err := snap.AllocByIDPrefix(args.AllocID)
```

Comment:
So as I mentioned earlier we don't want to put this logic here (for all the endpoints). These endpoint are for looking up the particular id.

I think the best way to add this feature is to add a `?prefix=true` parameter to the list requests (there is one for job/eval/alloc).

You would have to add:
* parsing [here](https://github.com/hashicorp/nomad/blob/master/command/agent/http.go#L271) and [example call](https://github.com/hashicorp/nomad/blob/master/command/agent/job_endpoint.go#L23)
* add to struct [here](https://github.com/hashicorp/nomad/blob/master/nomad/structs/structs.go#L58)
* update list methods to use the new state store calls you added. [Example](https://github.com/hashicorp/nomad/blob/master/nomad/job_endpoint.go#L260)
   * So if prefix is set use `JobByIdPrefix` instead of `Jobs`
* The CLI commands would then need to first call with the specific ID and if there is an error call list with the prefix and then show the output to the user.

This has the benefit that it makes the list endpoint more usable, maintains the API of the get endpoints, makes everything much more testable, and we aren't using the error as a return value.